internal/charmstore/addentity.go

```diff
@@ -111,7 +92,16 @@ func (s *Store) addEntityWithArchive(url *router.ResolvedURL, archive interface{
 //	params.ErrEntityIdNotAllowed if the id may not be created.
 //	params.ErrInvalidEntity if the provided blob is invalid.
 func (s *Store) UploadEntity(url *router.ResolvedURL, blob io.Reader, blobHash string, size int64) error {
-	logger.Infof("uploading entity with blob hash %v", blobHash)
+	// Strictly speaking these tests are redundant, because a ResolvedURL should
+	// always be canonical, but check just in case anyway, as this is
+	// final gateway before a potentially invalid url might be stored
+	// in the database.
+	if url.URL.User == "" {
+		return errgo.WithCausef(nil, params.ErrEntityIdNotAllowed, "entity id does not specify user")
```

Comment:
actually most error messages use "entity". maybe we should standardise on "charm or bundle" but I don't like it as it means we'll need to change all the error messages if/when we add another entity type.